desktop/libs/notebook/src/notebook/connectors/base.py

```diff
@@ -424,7 +432,13 @@ def get_api(request, snippet):
   else:
     interpreter = get_interpreter(connector_type=connector_name, user=request.user)
 
-  interface = interpreter['interface']
+  interface = snippet.get('interface')
+  if interface is None:
+    interface = interpreter['interface']
+
+  options = snippet.get('options')
+  if options is not None:
+    interpreter['options'] = options
```

Comment:
Confused: A bit skeptical on this, why are we prioritising snippet over interpreter here? Is there a way to depend on interpreter only for interface? else we need to be super sure for no regressions if we are depending on snippet for interface.

Reply:
In rdbms.py script file, get_api function is executed by get_db_component function.
https://github.com/cloudera/hue/blob/44e4238fcb50b9ea614d11f77728bb70e0d640c1/desktop/libs/indexer/src/indexer/indexers/rdbms.py#L103

As you can see, parameters of get_api function are 'request' and dictionary type data(this is snippet).
Variables like type, interface, options, query_server and name will be replaced by the value entered by the user.
```python
{ 'type': type, 
  'interface': interface, 
  'options': options, 
  'query_server': query_server, 
  'name': name
}
```
However, the parameters are not passed properly. namely, user entered data is not passed.
Even if you fill the UI text box with valid value, 'Test Connection' will be failed. and error message will appear.
```
Connection Failed: Error accessing the database: Notebook connector interface not recognized:
```
---
I Checked the hue runcperver log and was able to find the line below
```
# interpreter and interface, compute are vacant.
[15/Apr/2023 22:03:30 +0900] base         DEBUG    Selected interpreter  interface= compute=None
```

And check hive case.
```
[11/Apr/2023 18:38:06 +0900] base         DEBUG    Selected interpreter hive interface=hiveserver2 compute=None
```
Now it is clear that parameters are not passed or are abandoned in the middle.

https://github.com/cloudera/hue/blob/44e4238fcb50b9ea614d11f77728bb70e0d640c1/desktop/libs/notebook/src/notebook/connectors/base.py#L409-L427

Checking code avobe, interface variable is from interpreter dictionary but interpreter variable is not set in the use cases(https://github.com/cloudera/hue/pull/3268#issuecomment-1509508660)

Therefore, I prioritize snippet over interepreter. 

Thank you for the comment and if you have any confusion, please let me know.

Reply:
Thanks for explaining the change thoroughly!
What is the interface value you get with this change for your test cases?

Basically what is the log statement now when you run your test-cases? Like the one below but for your change.
```
# interpreter and interface, compute are vacant.
[15/Apr/2023 22:03:30 +0900] base         DEBUG    Selected interpreter  interface= compute=None 
```

I am interested to know which interface it chooses in later part of the code for your test-cases. Is it this? https://github.com/cloudera/hue/pull/3268/files#diff-833eafd5accb4467609950009daa4403e9cf8b8189076edb6589e172034554acR470

Reply:
Exactly, this is my log statement after the change.
```python
[26/Apr/2023 08:56:44 +0900] base         DEBUG    Selected interpreter  interface=rdbms compute=None
```
There is no interpreter, but interface type is set to 'rdbms' so that user can continue with importer feature.

Reply:
I see, its `rdbms` here.
@HalimKim - Then maybe we can try populating it here: https://github.com/cloudera/hue/pull/3268/files#diff-833eafd5accb4467609950009daa4403e9cf8b8189076edb6589e172034554acR391-R392 and this should allow us to not depend on snippet then because `interpreter['interface']` should have it then.

Can you give it a try?

Reply:
Just like the other interpreter dict elements which have type and interface.

Reply:
```python
...
elif connector_type == 'custom':
      interpreter = [{
        'name': 'custom',
        'type': 'custom',
        'interface': 'rdbms',
        'options': {},
        'is_sql': False
      }]
...
```
Thank you for giving me a good idea. but if we depend on interpreter, another problem will occurs.
There is no problem with Selecting MySQL, Oracle or Postgresql. However, When selecting JDBC driver, interface type should be 'jdbc' not 'rdbms'. 
Therefore, It is possible to fill in the blank of 'custom' type interpreter, but we can't help prioritzing snippet over interpreter.

Reply:
Ahh I see, then let's prioritise snippet over interpreter only for `jdbc` and `rdbms` interfaces.
Something like?

```python
...
interface = interpreter['interface']

if snippet.get('interface') in ('jdbc', 'rdbms'):
  interface = snippet.get('interface')
  options = snippet.get('options')

...
```

Reply:
I agree with adding a if statement but It is better to utilize 'custom' interpreter.
What i mean is that if statement has to check it is custom interpreter rather than check snippet's interface is 'jdbc' or 'rdbms'.

My suggestion is to keep 'custom' interpreter as dummy and reconstruct 'custom' interpreter by getting snippet value.
```python
...
elif connector_type == 'custom':
  interpreter = [{
    'name': 'custom',
    'type': 'custom',
    'interface': '',
    'options': {},
    'is_sql': False
  }]
...
```

```python
...
# reconstruct 'custom' interpreter.
if snippet.get('type') and snippet.get('type') == 'custom': 
  interface = snippet.get('interface') 
  interpreter['options'] = snippet.get('options')
...
```

Moreover, There is some case that we need to prioritize interpreter over snippet. In the hue.ini file, user can define interpreter like below. 
```ini
[[interpreters]]
...
 [[[mysql]]]
  name=MySql JDBC
  interface=jdbc
  options='{"url": "jdbc:mysql://localhost:3306/test", "driver": "com.mysql.jdbc.Driver", "user": "test", "password": "test"}'
...
``` 
In that case, it is likely to be malfunction with mysql interpreter. 
I've tested two use cases(https://github.com/cloudera/hue/pull/3268#issuecomment-1509508660) and one more case that defines MySql JDBC interpreter above.

Reply:
Yes, I've added my change to the code. thank you for your concern.